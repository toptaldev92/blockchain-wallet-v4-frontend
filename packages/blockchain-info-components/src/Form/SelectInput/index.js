import React from 'react'
import PropTypes from 'prop-types'
import { equals, prop } from 'ramda'

import SelectInput from './template.js'

class SelectInputContainer extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      value: this.props.value
    }

    this.handleChange = this.handleChange.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    if (!equals(this.props.value, nextProps.value)) {
      this.setState({ value: nextProps.value })
    }
  }

  handleChange (item) {
    const value = prop('value', item)

    this.setState({
      value
    })
    if (this.props.onChange) {
      this.props.onChange(value)
    }
  }

  render () {
    const {
      elements,
      label,
      searchEnabled,
      disabled,
      grouped,
      ...rest
    } = this.props
    const items = elements[0].items

    return (
      <SelectInput
        items={items}
        disabled={disabled}
        defaultItem={this.state.value}
        defaultDisplay={label}
        handleChange={this.handleChange}
        searchEnabled={this.props.searchEnabled}
        grouped={grouped}
        {...rest}
      />
    )
  }
}

SelectInputContainer.propTypes = {
  elements: PropTypes.arrayOf(
    PropTypes.shape({
      group: PropTypes.string.isRequired,
      items: PropTypes.array.isRequired
    })
  ).isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  searchEnabled: PropTypes.bool,
  opened: PropTypes.bool,
  disabled: PropTypes.bool,
  grouped: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.object
  ]).isRequired,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  templateDisplay: PropTypes.func,
  templateHeader: PropTypes.func,
  templateItem: PropTypes.func
}

SelectInputContainer.defaultProps = {
  label: 'Select a value',
  searchEnabled: true,
  opened: false,
  grouped: false,
  disabled: false
}

export default SelectInputContainer
