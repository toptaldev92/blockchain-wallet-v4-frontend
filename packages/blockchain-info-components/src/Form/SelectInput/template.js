import { assoc, equals, filter, flatten, head, path } from 'ramda'
import PropTypes from 'prop-types'
import React from 'react'
import Select, { components } from 'react-select'
import styled from 'styled-components'

const StyledSelect = styled(Select)`
  font-weight: 500;
  font-family: 'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto,
    Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  font-size: ${props => (props.fontSize === 'small' ? '14px' : '16px')};

  .bc__menu {
    background-color: ${props => props.theme.white};
  }

  .bc__menu-list {
    padding: 0;
  }

  .bc__group {
    padding-bottom: 0;
    > div:nth-child(2) {
      .bc__option {
        padding-top: 6px;
      }
    }
  }

  .bc__group-heading {
    font-weight: 500;
    margin-bottom: 0px;
    color: ${props => props.theme['gray-6']};
  }

  .bc__placeholder {
    color: ${props => props.theme.grey100};
  }

  .bc__control {
    box-shadow: none;
    color: ${props => props.theme['gray-6']};
    background-color: ${props => props.theme.white};
    cursor: pointer;
    min-height: ${props => props.height};
    border-radius: 4px;
    border: 1px solid ${props => props.theme[props.borderColor]};
    &:hover {
      border: 1px solid ${props => props.theme[props.borderColor]};
    }
    &:active {
      border: 1px solid ${props => props.theme[props.borderColor]};
      box-shadow: none;
    }
    &:disabled {
      cursor: not-allowed;
      background-color: ${props => props.theme['gray-1']};
    }
    .bc__value-container {
      overflow: hidden;
    }

    input {
      border: none !important;
    }
  }

  .bc__indicator-separator {
    display: none;
  }

  .bc__option {
    cursor: pointer;
    font-size: 14px;
    color: ${props => props.theme['gray-6']};
    background-color: ${props => props.theme.white};
    &.bc__option--is-focused {
      background-color: ${props => props.theme.white};
      color: ${props => props.theme.blue900};
    }
    &.bc__option--is-selected {
      color: ${props => props.theme.blue900};
      background-color: ${props => props.theme.white};
      &:hover {
        color: ${props => props.theme.blue900};
      }
      * {
        color: ${props => props.theme.blue900};
      }
    }
    &:hover {
      background-color: ${props => props.theme.white};
      * {
        color: ${props => props.theme.blue900};
      }
    }
    * {
      font-weight: 500;
      color: ${props => props.theme['gray-6']};
      transition: color 0.3s;
    }
    > div {
      font-size: 14px;
    }
  }

  .bc__single-value {
    color: ${props => props.theme['gray-6']};
  }
`

const Option = props => {
  const itemProps = assoc('text', props.label, props)
  return (
    <components.Option {...props}>
      {props.selectProps.templateItem
        ? props.selectProps.templateItem(itemProps)
        : props.children}
    </components.Option>
  )
}

const Control = props => {
  return props.selectProps.hideFocusedControl &&
    props.selectProps.menuIsOpen ? null : (
    <components.Control {...props} />
  )
}

const ValueContainer = ({ children, ...props }) => {
  const displayProps = assoc(
    'text',
    path(['selectProps', 'value', 'label'], props),
    assoc('value', path(['selectProps', 'value', 'value'], props), props)
  )
  return (
    <components.ValueContainer {...props}>
      {props.selectProps.templateDisplay
        ? props.selectProps.templateDisplay(displayProps, children)
        : children}
    </components.ValueContainer>
  )
}

const DropdownIndicator = props => {
  return props.selectProps.hideIndicator ? null : (
    <components.DropdownIndicator {...props} />
  )
}

const IndicatorSeparator = props => {
  return props.selectProps.hideIndicator ? null : (
    <components.IndicatorSeparator {...props} />
  )
}

const selectBorderColor = state => {
  switch (state) {
    case 'initial':
      return 'grey100'
    case 'invalid':
      return 'error'
    case 'valid':
      return 'success'
    default:
      return 'grey100'
  }
}

const SelectInput = props => {
  const {
    className,
    defaultDisplay,
    defaultItem,
    disabled,
    errorState,
    filterOption,
    getRef,
    grouped,
    handleChange,
    height,
    hideFocusedControl,
    hideIndicator,
    items,
    menuIsOpen,
    menuPlacement,
    onBlur,
    onFocus,
    onKeyDown,
    openMenuOnFocus,
    searchEnabled,
    templateDisplay,
    templateItem
  } = props
  const options = grouped
    ? items
    : items.map(item => ({ value: item.value, label: item.text }))
  const groupedOptions = grouped && flatten(options.map(o => o.options))
  const defaultValue = grouped
    ? head(filter(x => equals(x.value, defaultItem), groupedOptions))
    : head(filter(x => equals(x.value, defaultItem), options))
  const borderColor = selectBorderColor(errorState)

  return (
    <StyledSelect
      borderColor={borderColor}
      className={className}
      classNamePrefix='bc'
      components={{
        Option,
        ValueContainer,
        Control,
        DropdownIndicator,
        IndicatorSeparator
      }}
      height={height}
      filterOption={filterOption}
      hideFocusedControl={hideFocusedControl}
      hideIndicator={hideIndicator}
      isDisabled={disabled}
      isSearchable={searchEnabled}
      menuIsOpen={menuIsOpen}
      menuPlacement={menuPlacement}
      onBlur={onBlur}
      onChange={handleChange}
      onFocus={onFocus}
      onKeyDown={onKeyDown}
      openMenuOnFocus={openMenuOnFocus}
      options={options}
      placeholder={defaultDisplay}
      ref={getRef}
      templateDisplay={templateDisplay}
      templateItem={templateItem}
      value={defaultValue}
    />
  )
}

SelectInput.propTypes = {
  items: PropTypes.array.isRequired,
  selected: PropTypes.object,
  expanded: PropTypes.bool,
  searchEnabled: PropTypes.bool,
  menuIsOpen: PropTypes.bool,
  openMenuOnFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  errorState: PropTypes.string,
  handleChange: PropTypes.func,
  handleClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  templateItem: PropTypes.func,
  getRef: PropTypes.func,
  fontSize: PropTypes.string,
  filterOption: PropTypes.func
}

export default SelectInput
