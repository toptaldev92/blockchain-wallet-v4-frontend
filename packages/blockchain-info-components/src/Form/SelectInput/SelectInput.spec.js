import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json';

import SelectInput from './template';

describe('SelectInput component', () => {
  it('default renders correctly', () => {
    const items = [
      { text: 'test', value: 'value' }
    ];
    const component = shallow(<SelectInput items={items} display='inherit' />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
