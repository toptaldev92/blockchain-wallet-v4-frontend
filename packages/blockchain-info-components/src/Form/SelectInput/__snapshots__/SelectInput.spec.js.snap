// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput component default renders correctly 1`] = `
<styled.div>
  <styled.button>
    <styled.div>
      hello
    </styled.div>
  </styled.button>
  <Styled(Icon)
    name="down-arrow"
  />
  <styled.div>
    <styled.div
      key="0"
      onMouseDown={[Function]}
    >
      <styled.div>
        test
      </styled.div>
    </styled.div>
  </styled.div>
</styled.div>
`;
